File: https:/github.com/abhinavtripathi17/DSA-Leetcode.git/84-LargestRectangleInHistogram/84-LargestRectangleInHistogram.cpp
// Last updated: 8/4/2026, 7:23:01 PM
class Solution {
public:
    int largestRectangleArea(vector<int>& heights) {
        int n = heights.size();
        int maxarea = 0;

        vector<int>leftmin(n) , rightmin(n);
        stack<int>st;

        for(int i = 0 ; i < n ; i++){
            while(!st.empty() && heights[st.top()] >= heights[i]){
                st.pop();
            }

            leftmin[i] = st.empty() ? 0 : st.top()+1;

            st.push(i); 
        }
        while(!st.empty()){
            st.pop();
        }

        for(int i = n-1 ; i >= 0 ; i--){
            while(!st.empty() && heights[st.top()] >= heights[i]){
                st.pop();
            }

            rightmin[i] = st.empty() ? n-1 : st.top() - 1;

            st.push(i);
        }

        for(int i = 0 ; i < n ; i++){
            int width = rightmin[i] - leftmin[i] + 1;
            maxarea = max(maxarea , width * heights[i]);
        }

        return maxarea;
    }
};

File: https:/github.com/abhinavtripathi17/DSA-Leetcode.git/84-LargestRectangleInHistogram/84-LargestRectangleInHistogram.java
// Last updated: 8/4/2026, 7:23:07 PM
class Solution {
    public int largestRectangleArea(int[] heights) {
        int n = heights.length;
        Stack<Integer>st = new Stack<>();
        int max_ans = 0;
        
        for(int i = 0 ; i <= n ; i++){
            int curr;
            if(i == n){
                curr = 0;
            }
            else{
                curr = heights[i];
            }

            while(!st.empty() && heights[st.peek()] > curr){
                int height = st.pop();
                int width;
                if(st.empty()){
                    width = i;
                }
                else{
                    width = i - st.peek() - 1;
                }

                max_ans = Math.max(max_ans , heights[height] * width);
            }
            st.push(i);
        }
        return max_ans;
    }
}